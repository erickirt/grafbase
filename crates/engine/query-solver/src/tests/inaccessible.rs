use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/union-intersection/a")
    B @join__graph(name: "b", url: "http://localhost:4200/union-intersection/b")
}

type Book
    @join__type(graph: A)
    @join__type(graph: B)
{
    title: String!
}

type Movie
    @join__type(graph: B)
{
    title: String!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
{
    media: Media
    book: Media @join__field(graph: A, type: "Book") @join__field(graph: B, type: "Media")
    song: Media @join__field(graph: A)
    viewer: Viewer
}

type Song
    @join__type(graph: A)
{
    title: String!
}

type Viewer
    @join__type(graph: A)
    @join__type(graph: B)
{
    media: ViewerMedia
    book: ViewerMedia @join__field(graph: A, type: "Book") @join__field(graph: B, type: "ViewerMedia")
    song: ViewerMedia @join__field(graph: A)
}

union Media
    @join__type(graph: A)
    @join__type(graph: B)
    @join__unionMember(graph: A, member: "Book")
    @join__unionMember(graph: B, member: "Book")
    @join__unionMember(graph: A, member: "Song")
    @join__unionMember(graph: B, member: "Movie")
 = Book | Song | Movie

union ViewerMedia
    @join__type(graph: A)
    @join__type(graph: B)
    @join__unionMember(graph: A, member: "Book")
    @join__unionMember(graph: B, member: "Book")
    @join__unionMember(graph: A, member: "Song")
    @join__unionMember(graph: B, member: "Movie")
 = Book | Song | Movie
"#;

#[test]
fn inaccessible_field_for_one_subgraph() {
    // title is available in subgraph B not in A. Query.media is accessible in both.
    assert_solving_snapshots!(
        "inaccessible_field_for_one_subgraph",
        SCHEMA,
        r#"
        query {
          media {
            ... on Movie {
              title
            }
          }
        }
        "#
    );
}
