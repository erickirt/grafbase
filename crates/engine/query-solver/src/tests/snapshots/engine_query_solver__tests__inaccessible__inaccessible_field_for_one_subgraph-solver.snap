---
source: crates/engine/query-solver/src/tests/inaccessible.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Movie.title\", style=dashed ]\n    2 [ label = \"Root#b\", shape=parallelogram, color=dodgerblue, style=dashed ]\n    3 [ label = \"media#b\", shape=box, color=dodgerblue, style=dashed ]\n    4 [ label = \"title#b\", shape=box, color=dodgerblue, style=dashed ]\n    0 -> 2 [ label = <<b>1</b>>, color=royalblue,fontcolor=royalblue,style=dashed ]\n    2 -> 3 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    3 -> 4 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n    4 -> 1 [ label = \"\", color=royalblue,fontcolor=royalblue,style=dashed ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Movie.title", steiner=0 ]
    2 [ label = "Root#b", steiner=0 ]
    3 [ label = "media#b", steiner=0 ]
    4 [ label = "title#b", steiner=0 ]
    0 -> 2 [ cost=1, steiner=0]
    2 -> 3 [ cost=0, steiner=0]
    3 -> 4 [ cost=0, steiner=0]
    4 -> 1 [ cost=0, steiner=0]
}
